neuralmonkey/decoders/decoder.py

```diff
@@ -30,34 +28,24 @@
     "LSTM": tf.contrib.rnn.LSTMCell
 }
 
+# pylint: disable=invalid-name
+RNNFeedables = extend_namedtuple(
+    "DecoderFeedables",
+    DecoderFeedables,
+    [("prev_rnn_state", tf.Tensor),
+     ("prev_rnn_output", tf.Tensor),
+     ("prev_contexts", List[tf.Tensor])])
 
-# The LoopState is a structure that works with the tf.while_loop function
-# the decoder loop state stores all the information that is not invariant
-# for the decoder run.
-LoopState = NamedTuple("LoopState",
-                       [("step", tf.Tensor),  # 1D int, number of the step
-                        ("input_symbol", tf.Tensor),  # batch of ints to vocab
-                        ("train_inputs", Optional[tf.Tensor]),
-                        ("prev_rnn_state", tf.Tensor),
-                        ("prev_rnn_output", tf.Tensor),
-                        ("rnn_outputs", tf.TensorArray),
-                        ("prev_logits", tf.Tensor),
-                        ("logits", tf.TensorArray),
-                        ("prev_contexts", List[tf.Tensor]),
-                        ("mask", tf.TensorArray),  # float matrix, 0s and 1s
-                        ("finished", tf.Tensor),  # batch-sized, bool
-                        ("attention_loop_states", List[Any])])  # see att docs
-
-
-# pylint: disable=too-many-public-methods,too-many-instance-attributes
-# Big decoder cannot be simpler. Not sure if refactoring
-# it into smaller units would be helpful
-class Decoder(ModelPart):
-    """Decoder class.
-
-    A class that manages parts of the computation graph that are
-    used for the decoding.
-    """
+RNNHistories = extend_namedtuple(
+    "RNNHistories",
+    DecoderHistories,
+    [("attention_histories", List[Tuple])])  # AttentionLoopStateTA and kids
+# pylint: enable=invalid-name
+
+
+# pylint: disable=too-many-instance-attributes
+class Decoder(SequenceDecoder):
+    """A class managing parts of the computation graph used during decoding."""
```

Comment:
update docstring

Standard recurrent decoder with attention as defined in Bahdanau et al. (2015) and extended in Libovický et al. (2016)

plus arxiv odkazy (ten náš je cuni system for wmt16 mmmt+ape)

Reply:
Když už se tam tohle bude psát, tak by bylo fér poznamenat, že to používá Conditional GRU (to bych citoval buď to pdf z githubu nebo nematusí článek) a umožňuje posílat attention na vstup v dalším kroku tak, jak to dělá Show, Attend and Tell (Xu et al., 201??)